include/cxxeffect/task.hpp

```diff
@@ -1,136 +1,95 @@
 #pragma once
-
-#include <exception>
 #include <functional>
 #include <future>
-#include <memory>
-#include <variant>
+#include <type_traits>
 
 namespace eff {
-
-    template<typename A>
+    // This class represents a task that can be executed synchronously or
+    // asynchronously, and produces a value
+    template <class T>
     class task;
 
-    template<typename A>
-    class raw_task {
-        public:
-        virtual A unsafeRunSync() const = 0;
-        virtual std::future<A> unsafeRunAsync() const = 0;
-    };
-
-    template<typename A>
-    class raw_task_pure final : public raw_task<A> {
-        A value;
-        public:
-        raw_task_pure(A a) : value(a) {}
-
-        A unsafeRunSync() const {
-            return value;
-        }
-
-        std::future<A> unsafeRunAsync() const {
-            return std::async(std::launch::async, [*this]() { return value; });
-        }
-    };
-
-    template<typename A>
-    class raw_task_delay final : public raw_task<A> {
-        std::function<A ()> thunk;
-        public:
-        raw_task_delay(std::function<A ()> a) : thunk(a) {}
-
-        A unsafeRunSync() const {
-            return thunk();
-        }
-
-        std::future<A> unsafeRunAsync() const {
-            return std::async(std::launch::async, thunk);
-        }
+    namespace _internal {
```

Comment:
Is this standard? I remember there being some sort of standardized way to have a "private" namespace.

Reply:
There are anonymous namespaces (which are unnamed). Anonymous namespaces have internal linkage, but it doesn't make sense to use them for things that have to be templated (like the class I used to check whether or not something was a type. There are two things to note here:

- It's templated, but it doesn't actually define any functions, so it doesn't get compiled down to assembly. There's no risk of code bloat.
- Because it's templated, it has to go in the same header file as the one that defines `task`. This means that it'll end up being in the same translation unit as whatever uses `task`, which means the internal linkage doesn't actually make it private, since internal linkage is internal to a translation unit (rather than a file, as it's sometimes mistakenly described as being)

This problem is solved by C++20's module system, which it introduced to the language. Modules are better than header files, and they allow you to only export specific things from the module, however GCC and Clang are both still in the process of implementing modules, and they're not ready for production use.

As a result, I used `_internal`. Using an underscore before a name to mark it as internal is something common to a lot of libraries written in C++. `_internal` is safe to use, because it's not reserved by the standard. (The standard reserves names containing two consecutive underscores, so `__internal` is off the table, and it reserves names containing an underscore followed by a capital letter, so `_Internal` is also off the table, but `_internal` is fine. See: <https://stackoverflow.com/a/228797>, which references the relevant portion of the C++ standard)